drivers/clock_control/clock_control_renesas_ra_cgc.c

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2024 Renesas Electronics Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <string.h>
+
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/kernel.h>
+#include <soc.h>
+#include <zephyr/dt-bindings/clock/ra_clock.h>
+#include <zephyr/drivers/clock_control/renesas_ra_cgc.h>
+
+static int clock_control_renesas_ra_on(const struct device *dev, clock_control_subsys_t sys)
+{
+	struct clock_control_ra_subsys_cfg *subsys_clk = (struct clock_control_ra_subsys_cfg *)sys;
```

Comment:
Please do not cast `clock_control_subsys`_t to different types ​​for on, off, and get_rate. For places where the same type is used as an argument in the clock_control API, please use a consistent type according to the definition of the original API.

Reply:
@soburi , I refer to some other clock control driver, they also do casting for subsys clock.

e.g. https://github.com/zephyrproject-rtos/zephyr/blob/fd0e46506fabd4c0c91ecb336a39cb6a6aa2b270/drivers/clock_control/clock_stm32_ll_u5.c#L151C9-L151C22

Please correct me if I'm wrong, as my understanding in the description in clock_control.h, the clock_control_subsys_t is designed to be casted to the data structure of sub-system driver instance being used.

/**
 * clock_control_subsys_t is a type to identify a clock controller sub-system.
 * Such data pointed is opaque and relevant only to the clock controller
 * driver instance being used.
 */
typedef void *clock_control_subsys_t;

Reply:
Sorry, I checked something incorrectly, so I'll comment again.

Here, casting itself is not a problem.
The type is defined as void*, so it can be any type.

By the way, how is the clock_control_subsys_t argument processed in get_rate?

Here, the function is expected to be called with the same clock_control_subsys_t value as in the on/off case, so the corresponding clock frequency is expected to be obtained.

In other words, if the same value as a clock_control_subsys_t that allows on and off to operate normally is passed to get_rate, that frequency is expected to be obtained for this. We think it should be possible to get the pclk[abc...] value corresponding to each peripheral, not the system clock.

Reply:
@soburi -san, this implement treat each pclk as a clock device, hence it will get the rate of pclk[abc...] by the device config. Similar as the implementation of fixed clock driver in Zephyr

Reply:
In Zephyr, it is expected to call the `get_rate` of clock_controller with the value of `clocks` for each node set in the device tree to obtain the clock frequency required by that peripheral. The API of clock_controller is designed with this usage in mind.

In the case of Renesas RA, the FSP mechanism can be used to obtain the clock information required by the driver without using this method,  so this is not a problem at a glance, but this is the FSP way and not Zephyr's. Since clocks affect the entire system, they should be designed appropriately according to Zephyr's custom; the get_rate of clock_controller should be implemented appropriately.

For example, the SCI uses pclka in RA4M1 and the pclkb in RA2A1.
I hope it can switch between these in the device tree description and obtain the frequency of each clock through get_rate.

In practice, I think adding one cell to specify the type of clock is enough.